### PR TITLE
Change default feature set to MVP

### PR DESCRIPTION
**Pull request by tlively**
This greatly simplifies the feature handling logic and does not
change the user experience when a target features section is present.
Also refactors features to be stored on the Module after they are read
out of the target features section, allowing us to emit a (potentially
modified) target features section and conditionally emit sections
based on features, such as the DataCount section.

### CODE REVIEW

File: test/debugInfo.fromasm.clamp.no-opts.map [hunk 1]
```diff
@@ -1 +1 @@
-{"version":3,"sources":["tests/hello_world.c","tests/other_file.cpp","return.cpp","even-opted.cpp","fib.c","/tmp/emscripten_test_binaryen2_28hnAe/src.c","(unknown)"],"names":[],"mappings":"yLAIA,IACA,ICyylTA,aC7vlTA,OAkDA,0BCnGA,OACA,OACA,uBCAA,4BAKA,QAJA,OADA,8CAKA,0ICsi1DA,MCrvyDA"}
```
Comment (tlively) [inline test/debugInfo.fromasm.clamp.no-opts.map]: TBH I have no idea what this change means.

File: check.py [hunk 2]
```diff
@@ -291,7 +290,7 @@ def run_wasm_reduce_tests():
     before = os.stat('a.wasm').st_size
     run_command(WASM_REDUCE + ['a.wasm', '--command=%s b.wasm --fuzz-exec' % WASM_OPT[0], '-t', 'b.wasm', '-w', 'c.wasm'])
     after = os.stat('c.wasm').st_size
-    assert after < 0.6 * before, [before, after]
+    assert after < 0.7 * before, [before, after]
```
Comment (tlively) [inline check.py]: I haven't really examined why this changed yet.

File: src/wasm.h [hunk 3]
```diff
@@ -924,6 +924,9 @@ class Module {
   std::vector<UserSection> userSections;
   std::vector<std::string> debugInfoFileNames;
 
+  FeatureSet features = FeatureSet::MVP;
+  bool emitFeaturesSection = false;
```
Comment (kripken) [inline src/wasm.h]: Perhaps we should handle the features section the same way we handle the names section - we read it if it's there, but we don't emit it unless the user specifically asks us to?

File: src/wasm/wasm-binary.cpp [hunk 4]
```diff
@@ -486,14 +489,49 @@ void WasmBinaryWriter::writeLateUserSections() {
...
+  auto start = startSection(BinaryConsts::User);
   writeInlineString(section.name.c_str());
   for (size_t i = 0; i < section.data.size(); i++) {
     o << uint8_t(section.data[i]);
   }
   finishSection(start);
 }
 
+void WasmBinaryWriter::writeFeaturesSection() {
+  if (!wasm->emitFeaturesSection || wasm->features <= FeatureSet{}) {
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: does the right side of the if mean "if MVP, don't emit the features section"?

File: src/wasm/wasm-binary.cpp [hunk 5]
```diff
@@ -486,14 +489,49 @@ void WasmBinaryWriter::writeLateUserSections() {
...
+    return;
+  }
+
+  // TODO(tlively): unify feature names with rest of toolchain and use
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: i don't understand this TODO - are these not already in harmony with the rest of the toolchain? What would need to change?

File: test/debugInfo.fromasm.clamp.no-opts.map (hunk 1 above)
Comment (kripken) [reply test/debugInfo.fromasm.clamp.no-opts.map]: I think it means that things moved around - the location of code is not the same as before. That's surprising, I think?

File: src/wasm.h (hunk 3 above)
Comment (tlively) [reply src/wasm.h]: Hmm, it would make sense to unify what we do with the features section and the names section (and the producers section? I forget what we do there). OTOH, it is very useful to preserve the features section through long chains of wasm-opt calls such as when running the reducer. Otherwise all but the first call of wasm-opt would have a validation failure.

File: src/wasm/wasm-binary.cpp (hunk 4 above)
Comment (tlively) [reply src/wasm/wasm-binary.cpp]: Yes, I forgot about the `isMVP()` method which I should definitely use here.

File: src/wasm/wasm-binary.cpp (hunk 5 above)
Comment (tlively) [reply src/wasm/wasm-binary.cpp]: the names used here are different from the names in wasm-features.h. For example "threads" is called "atomics" in the rest of the toolchain. There are other differences as well.

File: src/wasm.h (hunk 3 above)
Comment (tlively) [reply src/wasm.h]: I believe we decided offline to leave this how it is for now but to leave this open for unification with how we treat other sections in the future.

Comment (kripken) [reply src/wasm.h]: (and rename to `hasFeaturesSection`)

Comment (tlively) [reply src/wasm.h]: Right, thank you.

File: test/debugInfo.fromasm.clamp.no-opts.map (hunk 1 above)
Comment (tlively) [reply test/debugInfo.fromasm.clamp.no-opts.map]: Aha, this is because I changed the element section to only be emitted if it has any elements. That moved the code section up in the file.

**Comment by kripken** [comment]
Having to sprinkle `-all` everywhere (to tell it "all features are ok") worries me that it will be annoying in the day to day. How do other tools do this, like wabt? cc @binji

**Comment by tlively** [comment]
I believe WABT simply enables all features all the time (AFAIK it doesn't have any feature options). This makes sense for tools that read but do not modify modules and is what we do for wasm-as and wasm-dis. Tools that can modify modules need to know what features they are allowed to use, and it seems best to be conservative about that.

That being said, we probably do have more `-all` sprinkled around than would be optimal. In check.py there is one instance where we pass `-all` to wasm-shell. wasm-shell should be changed to assume `-all`. But all other instances of `-all` are passed to wasm-opt, which I think makes sense.

**Comment by tlively** [comment]
@kripken PTAL. This PR move towards implementing the plan described in #2004. I also expect this to fix current waterfall breakage, though I have not finished testing locally.

 - MVP is the default feature set.
 - When there is a target features section and no command line feature flags, the features section is used
 - When there is a target features section and there are command line feature flags, error if they do not exactly match, unless --detect-features was one of the feature flags.
 - --detect-features sets the current feature set equal to the contents of the target features section

There is not yet a way to crawl a binary and detect what features it uses.

File: src/tools/wasm-reduce.cpp [hunk 6]
```diff
@@ -321,6 +321,7 @@ struct Reducer : public WalkerPass<PostWalker<Reducer, UnifiedExpressionVisitor<
     Module wasm;
     ModuleReader reader;
     reader.read(working, *module);
+    wasm.features = FeatureSet::All;
```
Comment (kripken) [inline src/tools/wasm-reduce.cpp]: I'm not sure this is optimal - it might lead the optimizer to attempt odd things that hurt reduction. Please add a comment that we should likely autodetect features from the binary here instead, once we can do that.

File: src/tools/wasm-reduce.cpp [hunk 7]
```diff
@@ -999,7 +1000,7 @@ int main(int argc, const char* argv[]) {
   std::cerr << "|checking that command has expected behavior on canonicalized (read-written) binary\n";
   {
     // read and write it
-    auto cmd = Path::getBinaryenBinaryTool("wasm-opt") + " " + input + " -o " + test;
+    auto cmd = Path::getBinaryenBinaryTool("wasm-opt") + " " + input + " -all -o " + test;
```
Comment (kripken) [inline src/tools/wasm-reduce.cpp]: same issue

File: src/wasm/wasm-binary.cpp [hunk 8]
```diff
@@ -723,8 +723,8 @@ void WasmBinaryBuilder::readUserSection(size_t payloadLen) {
 
 uint8_t WasmBinaryBuilder::getInt8() {
   if (!more()) throwError("unexpected end of input");
-  if (debug) std::cerr << "getInt8: " << (int)(uint8_t)input[pos] << " (at " << pos << ")" << std::endl;
-  return input[pos++];
+  if (debug) std::cerr << "getInt8: " << (int)(uint8_t)input.at(pos) << " (at " << pos << ")" << std::endl;
+  return input.at(pos++);
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: why is this change here? seems offtopic for the PR; also we already have a check right above it that throws an error.

Comment (tlively) [reply src/wasm/wasm-binary.cpp]: Right, sorry. My reverting of this change in #2001 was not backported to this branch. Fixing.

File: CHANGELOG.md [hunk 9]
```diff
@@ -29,6 +29,8 @@ v78
 - Add `segmentPassive` argument to `BinaryenSetMemory` for marking segments
   passive.
 - Make `-o -` print to stdout instead of a file named "-".
+- Change default feature set in the absence of a target features section from
```
Comment (sbc100) [inline CHANGELOG.md]: This should go up above in `Current Trunk`

Comment (tlively) [reply CHANGELOG.md]: Thanks.

File: test/unit/utils.py [hunk 10]
```diff
@@ -14,3 +14,22 @@ def roundtrip(testcase, filename, opts=[]):
   testcase.assertEqual(p.stderr, '')
   with open(path, 'rb') as f:
     testcase.assertEqual(str(p.stdout), str(f.read()))
+
+
+def disassemble(self, filename):
```
Comment (sbc100) [inline test/unit/utils.py]: Surely these should be left as members?   They take `self` as arg0 after all.

Comment (tlively) [reply test/unit/utils.py]: Yes, but I want to share them across multiple `unittest.TestCase` subclasses. I suppose the proper way to do that would be to make all my test cases inherit from a new class that contains these methods.

**Comment by binji** [comment]
> Having to sprinkle -all everywhere (to tell it "all features are ok") worries me that it will be annoying in the day to day. How do other tools do this, like wabt? cc @binji

> I believe WABT simply enables all features all the time (AFAIK it doesn't have any feature options). 

Sorry, missed this. Wabt does have feature flags for all tools. But `wasm-objdump` enables all features by default.
